### ISSUE TITLE
Better documentation for tf.extract_image_patches

### ISSUE THREAD

**Issue by bodokaiser** [tensorflow/tensorflow#6847]
In this [issue](https://github.com/tensorflow/tensorflow/issues/6743#issuecomment-272525783) someone requested a reverse operation to `tf.extract_image_patches`. The comments suggest that there exists a [gradient operation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/array_grad.py#L575) for this purpose which was added with [this PR](https://github.com/tensorflow/tensorflow/pull/3672).

Unfortunately it is not obvious how to apply this gradient operation and there are no information on this in the [api docs](https://www.tensorflow.org/versions/master/api_docs/python/array_ops/slicing_and_joining#extract_image_patches) therefor I request to add an example to [tf.extract_image_patches](https://github.com/tensorflow/tensorflow/blob/a4c8df209d7413068f4ed3e71c43eb798fbd5580/tensorflow/g3doc/api_docs/python/functions_and_classes/shard4/tf.extract_image_patches.md) where one transforms an `image` tensor of shape `(image_height, image_width, channels)` to `(patch_num, patch_height, patch_width, channels)` and vice versa.

Thank you in advance!

**Comment by jart** [comment]
@dr4b @vrv Want to bring to your attention that two users have requested better documentation for ExtractImagePatches, which currently has a one-liner doc. I think more accessible documentation is always a good thing. I've marked it "contributions welcome" but feel free to self-assign if it's something you're interested in doing.

**Comment by gunan** [comment]
@wolffg This still seems to be an issue. Could you take a look?

**Comment by agniszczotka** [comment]
@wolffg Is it possible to use tf.extract_image_patches as a not-trainable 2d convolution layer? how will the loss be computed with mixed tf.extract_image_patches and 2d convolutions?

**Comment by wolffg** [comment]
I think @vrv is more likely to be able to clear this up than I.

**Comment by vrv** [comment]
@agniszczotka: if you don't want to have backprop through extract_image_patches, wrap the output of tf.extract_image_patches() in tf.stop_gradient().

I think there's two issues being conflated here:
1) extract_image_patches doesn't have good documentation, and should be improved.  I would love one of the authors or maintainers of this code to add some.  @gpapan and @benoitsteiner ?

2) It sounds like people want access to the gradient function of tf.extract_image_patches() without having to use it as a gradient (essentially a reduction) as part of the API with corresponding documentation.  This is similar to people who want 'deconvolution' (the backward pass of convolution) as its own API.  Such a feature is possible just like with deconvolution, it just requires writing the code and exposing it in the API after getting API approval.

**Comment by MarvinTeichmann** [comment]
@vrv  "2. It sounds like people want access to the gradient function of tf.extract_image_patches() without having to use it as a gradient (essentially a reduction). This is similar to people who want 'deconvolution' (the backward pass of convolution) as its own API. "

This is a quite common use-case for many ops even  beyond 'deconvolution'  and 'extract_image_patches()'. Do you think it would be possible to great a general API, which lets you (easily) evaluate any op  (or even subgraph) backwards? 

The api could work similar to pytorchs `.backward()` call which can be applied to subgraphs.

**Comment by vrv** [comment]
@alextp What do you think?

The equivalent of .backward() might be exposing ops.get_gradient_function(), and would provide a mechanism to get the gradient of an op, but not as part of a documented API.  Perhaps that is good enough though?

**Comment by alextp** [comment]
You can get the equivalent of .backward() now for any tf op by doing something like tfe.gradients_function(tf.extract_image_patches)(arguments_to_image_patches ,..., dy=arguments_to_the_gradient_function) if tf.extract_image_patches has a gradient defined. When executing graphs this will only compute the forward pass if the backward function needs it (like how the gradient of exp or the gradient of relu uses the output of relu).

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by dynamicwebpaige** [comment]
Closing this issue, as `tf.extract_image_patches` is no longer supported in TensorFlow 2.0.

**Comment by mavenlin** [comment]
@dynamicwebpaige The reason for closing it is not valid because `tf.extract_image_patches` is not removed but just renamed to `tf.image.extract_patches`. I suggest edit the tile to "reverse op for `tf.image.extract_patches`" rather than closing it. This is not solved yet.
